### PR TITLE
Removed react-native-router-flux and added custom navigation logic

### PR DESCRIPTION
**Pull request by hmhealey**
The basic way this works now is that we have the navigation section of the redux store that's controlled with standard actions and a reducer. Right now, it only uses simple actions that push, pop, and reset the stack, but it has support for anything that NavigationExperimental can handle.

Overview of directory changes:
- src/actions/navigation - For simple navigation actions that can be reused. It might make sense to move these into view actions in the future since while they seem reusable now, they'll become more view specific once we tighten up the app navigation fow
- src/navigation
    - .../router.js - This is what uses the navigation state to render the whole app using NavigationExperimental. It also renders things like the navigation bar and transitions.
    - .../routes.js - These are plain objects representing each route in the app. It can also store properties of that route such as whether it has a title, back button, how it transitions, etc
-src/scenes - This used to be named src/routes, but I think it was a good idea to separate the definition of a route (which usually involves how you get there) from a scene (which is the actual contents of the route) and to make it more clear that this navigation isn't using a router with a path like we had before.
    - .../index.js - This provides a function mapping route keys to the scene that they use. This could've just been a field in src/navigation/routes.js, but I wanted to keep them separate to simplify dependencies because the actions and reducer for navigation both import routes.js

### CODE REVIEW

File: src/navigation/router.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        //     // Override modals to come in vertically
+        //     cardStyle = NavigationCard.CardStackStyleInterpolator.forVertical(props);
+        // }
+
+        // let panHandlers;
+        // if (currentSceneIsModal) {
+        //     // Explicitly set pan handlers to null to let NavigationCard override it and disable swiping back to
+        //     // close the modal
+        //     panHandlers = null;
+        // }
```
Comment (enahum) [inline src/navigation/router.js]: are these lines commented on purpose?

File: src/actions/navigation/index.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+export function goToSelectTeam() {
+    return async (dispatch, getState) => {
+        dispatch({
+            type: NavigationTypes.NAVIGATION_PUSH,
+            route: Routes.SelectTeam
+        }, getState);
+    };
+}
+
+export function goToChannel() {
```
Comment (enahum) [inline src/actions/navigation/index.js]: If i understand correctly we should pass here the channel to which we want to switch, right?

File: src/components/logout/logout_container.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2016 Mattermost, Inc. All Rights Reserved.
```
Comment (enahum) [inline src/components/logout/logout_container.js]: Why are you calling this file `logout_container` and other containers are just the index file of the folder? maybe we should rename this to be `/src/components/logout/index.js`

File: src/actions/views/root.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2016 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import {NavigationTypes} from 'constants';
+import Routes from 'navigation/routes';
+
```
Comment (enahum) [inline src/actions/views/root.js]: Why is this here and not in the navigation actions?

File: src/layouts/root_layout/root_layout_container.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import {connect} from 'react-redux';
+
+import Config from 'config';
+
+import RootLayout from './root_layout';
+
+function mapStateToProps(state, ownProps) {
+    let locale = Config.DefaultLocale;
+    if (state.entities.users.currentId && state.entities.users.profiles[state.entities.users.currentId]) {
+        locale = state.entities.users.profiles[state.entities.users.currentId].locale;
```
Comment (enahum) [inline src/layouts/root_layout/root_layout_container.js]: try and avoid multiple of this `state.entities.users.profiles[state.entities.users.currentId]`

if you can do something like

```javascript
const users = state.entities.users;
const currentId = users.currentId;
const user = users.profiles[currentId]
```

the differences is not much in terms of performance but it helps a lot when reading

File: src/utils/users.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2016 Mattermost, Inc. All Rights Reserved.
```
Comment (enahum) [inline src/utils/users.js]: This file should be deleted

**Comment by enahum** [comment]
@hmhealey where are you with this changes? I would love to merge this tomorrow morning

File: src/components/logout/logout_container.js (hunk 3 above)
Comment (hmhealey) [inline src/components/logout/logout_container.js]: I wanted to make it easier to find the files. We already have half a dozen `channel.js` files and I didn't want to add more confusion to that by having half these files called `index.js`. I can add a separate index.js though so we ca get the nice `import Logout from 'components/logout';` without having to specify the full path.

Comment (enahum) [inline src/components/logout/logout_container.js]: That's what im saying perhaps each one of the components that are in a folder should have an index.js file to accomplish exactly the same thing, don't you think?

File: src/actions/views/root.js (hunk 4 above)
Comment (hmhealey) [inline src/actions/views/root.js]: I'm actually thinking of eventually moving all of these into view action creators and not having a separate path for navigation ones. With the navigation being a stack, each navigation action creator will be likely depend heavily on which view in. We just need a NAVIGATION_PUSH to get from SelectServer to Login, but to get from SelectTeam to Login, we need a NAVIGATION_POP.

File: src/layouts/root_layout/root_layout_container.js (hunk 5 above)
Comment (hmhealey) [inline src/layouts/root_layout/root_layout_container.js]: Will do. Selectors will help this when I get those done.

File: src/actions/navigation/index.js (hunk 2 above)
Comment (hmhealey) [inline src/actions/navigation/index.js]: Right now, it's just switching to the scene named Channel, not to a particular channel. I'll leave it as it is for now, but I'll try to come up with a better name for it. We don't really have a better name for the main view/center channel/channel view right now, but we do need one to avoid ambiguity

File: src/components/logout/logout_container.js (hunk 3 above)
Comment (hmhealey) [inline src/components/logout/logout_container.js]: Done. Now, each component with a controller should look like
```
components/
    button/
        button.js
        button_container.js
        index.js
```
so we can always import it as `components/button` regardless of whether or not it has a container.

File: src/utils/users.js (hunk 6 above)
Comment (enahum) [inline src/utils/users.js]: Can you please delete this file, its here cause of a failed rebase, after that we can merge this
